crates/primitives/starknet/src/transaction/execution.rs

```diff
@@ -0,0 +1,566 @@
+        // Handle nonce.
+
+        if !disable_nonce_validation {
+            Self::handle_nonce(account_tx_context, state)?;
+        }
+
+        // Check fee balance.
+        if account_tx_context.max_fee != Fee(0) {
+            let (balance_low, balance_high) =
+                state.get_fee_token_balance(block_context, &account_tx_context.sender_address)?;
+            // TODO(Dori, 1/7/2023): If and when Fees can be more than 128 bit integers, this check
+            //   should be updated.
+            if balance_high == StarkFelt::from(0_u8) && balance_low < StarkFelt::from(account_tx_context.max_fee.0) {
+                return Err(TransactionExecutionError::MaxFeeExceedsBalance {
+                    max_fee: account_tx_context.max_fee,
+                    balance_low,
+                    balance_high,
+                });
+            }
+        }
+
+        Ok(())
+    }
+
+    fn execute<S: State + StateChanges + FeeConfig>(
+        &self,
+        state: &mut S,
+        block_context: &BlockContext,
+        is_query: bool,
+        disable_nonce_validation: bool,
+    ) -> TransactionExecutionResult<TransactionExecutionInfo> {
+        let mut execution_resources = ExecutionResources::default();
+        let mut remaining_gas = TX_INITIAL_AVAILABLE_GAS;
+
+        let account_tx_context = self.get_account_transaction_context(is_query);
+
+        // Nonce and fee check should be done before running user code.
+        Self::handle_nonce_and_check_fee_balance(state, block_context, &account_tx_context, disable_nonce_validation)?;
+
+        // execute
+        let ValidateExecuteCallInfo { validate_call_info, execute_call_info, revert_error } = self.execute_inner(
+            state,
+            block_context,
+            &mut execution_resources,
+            &mut remaining_gas,
+            &account_tx_context,
+        )?;
+
+        let (actual_fee, fee_transfer_call_info, actual_resources) = self.handle_fee(
+            state,
+            &execute_call_info,
+            &validate_call_info,
+            &mut execution_resources,
+            block_context,
+            account_tx_context,
+        )?;
+
+        let tx_execution_info = TransactionExecutionInfo {
+            validate_call_info,
+            execute_call_info,
+            fee_transfer_call_info,
+            actual_fee,
+            actual_resources,
+            revert_error,
+        };
+
+        Ok(tx_execution_info)
+    }
+
+    fn handle_fee<S: State + StateChanges + FeeConfig>(
+        &self,
+        state: &mut S,
+        execute_call_info: &Option<CallInfo>,
+        validate_call_info: &Option<CallInfo>,
+        execution_resources: &mut ExecutionResources,
+        block_context: &BlockContext,
+        account_tx_context: AccountTransactionContext,
+    ) -> TransactionExecutionResult<(Fee, Option<CallInfo>, ResourcesMapping)> {
+        let actual_resources = compute_transaction_resources(
```

Comment:
In 0.12.1 the blockifier also takes into account reverted steps when calculating transaction resources, I guess we are missing that part here - https://github.com/starkware-libs/blockifier/blob/e8ee613d05cf9c55bdb10b871b59365fca04b6ee/crates/blockifier/src/transaction/account_transaction.rs#L706

If that is the case, I am not sure if we really need to fix that in this PR as this PR is already big, maybe we can have a new issue to have test cases for reverted txns and also for this logic

Reply:
Let's do it in another issue then

Reply:
could you create it pls @apoorvsadana ?

Reply:
Yep, will do it